src/key/opaque_key_nostd.rs

```diff
@@ -0,0 +1,168 @@
+
+use alloc::String;
+use alloc::borrow::Cow;
+use alloc::Clone;
+
+use core::convert::{From,Into,AsRef};
+use core::ops::Deref;
+use core::str::FromStr;
+use core::cmp::PartialEq;
+use core::hash::{Hash,Hasher};
+use core::iter::{FromIterator, IntoIterator};
+use core::fmt;
+
+/// Opaque Key is a representation of a key.
+///
+/// It is owned, and largely forms a contract for
+/// key to follow.
+pub struct Key {
+    data: Cow<'static, str>
+}
+impl Key {
+
+    pub fn as_str<'a>(&'a self) -> &'a str {
+        self.data.as_ref()
+    }
+
+    pub fn into_string(&self) -> String {
+        match &self.data {
+            &Cow::Borrowed(ptr) => String::from(ptr),
+            &Cow::Owned(ref ptr) => ptr.clone()
+        }
+    }
+
+    pub fn into_owned(&self) -> String {
+        match &self.data {
+            &Cow::Borrowed(ptr) => String::from(ptr),
+            &Cow::Owned(ref ptr) => ptr.clone()
+        }
+    }
+}
+impl Default for Key {
+    fn default() -> Key {
+        Key {
+            data: Cow::Borrowed("")
+        }
+    }
+}
+impl Hash for Key {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self.data {
+            Cow::Borrowed(ref ptr) => ptr.hash(state),
+            Cow::Owned(ref ptr) => ptr.hash(state)
+        }
+    }
+}
+impl Clone for Key {
+    fn clone(&self) -> Key {
+        match self.data {
+            Cow::Borrowed(ptr) => Key::from(ptr)
+            Cow::Owned(ref ptr) => Key::from(ptr.clone())
+        }
+    }
+}
+impl From<&'static str> for Key {
+    #[inline(always)]
+    fn from(data: &'static str) -> Key {
+        Key {
+            data: Cow::Borrowed(data)
+        }
+    }
+}
+impl From<String> for Key {
+    #[inline(always)]
+    fn from(data: String) -> Key {
+        Key{
+            data: Cow::Owned(data)
+        }
+    }
+}
+impl Into<String> for Key {
```

Comment:
I thought "Library authors should not directly implement this trait, but should prefer implementing the From trait (...)" from https://doc.rust-lang.org/std/convert/trait.Into.html. So should this be `impl From<Key> for String` ?